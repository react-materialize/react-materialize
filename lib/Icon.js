'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _constants = require('./constants');

var _constants2 = _interopRequireDefault(_constants);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Icon = function (_React$Component) {
  _inherits(Icon, _React$Component);

  function Icon() {
    _classCallCheck(this, Icon);

    return _possibleConstructorReturn(this, (Icon.__proto__ || Object.getPrototypeOf(Icon)).apply(this, arguments));
  }

  _createClass(Icon, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var classes = {
        'material-icons': true
      };
      _constants2.default.PLACEMENTS.forEach(function (p) {
        classes[p] = _this2.props[p];
      });
      _constants2.default.ICON_SIZES.forEach(function (s) {
        classes[s] = _this2.props[s];
      });
      return _react2.default.createElement(
        'i',
        { className: (0, _classnames2.default)(classes, this.props.className) },
        this.props.children
      );
    }
  }]);

  return Icon;
}(_react2.default.Component);

Icon.propTypes = {
  /*
   * Classname passed to i tag
   */
  className: _propTypes2.default.string,
  /*
   * Icon type: <a href='https://material.io/icons/'>https://material.io/icons/</a>
   */
  children: _propTypes2.default.string,
  /*
   * Placement for icon if used beside a text ↓
   */
  left: _propTypes2.default.bool,
  center: _propTypes2.default.bool,
  right: _propTypes2.default.bool,
  /*
   * Sizes for icons ↓
   */
  tiny: _propTypes2.default.bool,
  small: _propTypes2.default.bool,
  medium: _propTypes2.default.bool,
  large: _propTypes2.default.bool
};

exports.default = Icon;