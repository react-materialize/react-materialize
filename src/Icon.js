import React from 'react';
import PropTypes from 'prop-types';
import constants from './constants';
import cx from 'classnames';

class Icon extends React.Component {
  render() {
    const classes = {
      'material-icons': true
    };
    constants.PLACEMENTS.forEach(p => {
      classes[p] = this.props[p];
    });
    constants.ICON_SIZES.forEach(s => {
      classes[s] = this.props[s];
    });
    return (
      <i className={cx(classes, this.props.className)}>{this.props.children}</i>
    );
  }
}

Icon.propTypes = {
  /*
   * Classname passed to i tag
   */
  className: PropTypes.string,
  /*
   * Icon type: <a href='https://material.io/icons/'>https://material.io/icons/</a>
   */
  children: PropTypes.string,
  /*
   * Placement for icon if used beside a text ↓
   */
  left: PropTypes.bool,
  center: PropTypes.bool,
  right: PropTypes.bool,
  /*
   * Sizes for icons ↓
   */
  tiny: PropTypes.bool,
  small: PropTypes.bool,
  medium: PropTypes.bool,
  large: PropTypes.bool
};

export default Icon;
