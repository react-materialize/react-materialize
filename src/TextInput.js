import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import idgen from './idgen';
import constants from './constants';

class TextInput extends Component {
  constructor(props) {
    super(props);

    this.id = props.id || idgen();

    if (props.password) {
      this.id = 'password';
    }

    if (props.email) {
      this.id = 'email';
    }
  }

  componentDidUpdate(prevProps) {
    const { value } = this.props;

    if (value !== prevProps.value) {
      M.updateTextFields();
    }
  }

  render() {
    const {
      s,
      m,
      l,
      xl,
      disabled,
      noLayout,
      placeholder,
      icon,
      label,
      inputClassName,
      success,
      error,
      password,
      email,
      validate,
      defaultValue,
      value,
      type,
      ...other
    } = this.props;

    const sizes = { s, m, l, xl };

    let responsiveClasses;
    if (!noLayout) {
      responsiveClasses = { col: true };
      constants.SIZES.forEach(size => {
        responsiveClasses[size + sizes[size]] = sizes[size];
      });
    }

    const wrapperClasses = cx('input-field', responsiveClasses);

    let computedType;
    if (type) {
      computedType = type;
    } else if (password) {
      computedType = 'password';
    } else if (email) {
      computedType = 'email';
    } else {
      computedType = 'text';
    }

    const inputProps = {
      placeholder,
      type: computedType,
      id: this.id,
      value: value,
      defaultValue: defaultValue,
      disabled,
      ...other
    };

    const renderLabel = () =>
      label && (
        <label
          className={cx({ active: value || placeholder })}
          data-success={success}
          data-error={error}
          htmlFor={inputProps.id}
        >
          {label}
        </label>
      );

    const renderHelper = () =>
      [error || success] && (
        <span
          className="helper-text"
          data-error={error}
          data-success={success}
        />
      );

    const renderIcon = () =>
      icon && <i className="material-icons prefix">{icon}</i>;

    return (
      <div className={wrapperClasses}>
        {renderIcon()}
        <input
          ref={el => {
            this.inputRef = el;
          }}
          className={cx({ validate }, inputClassName)}
          {...inputProps}
        />
        {renderLabel()}
        {renderHelper()}
      </div>
    );
  }
}

TextInput.propTypes = {
  /*
   * Strip away all layout classes such as col and sX
   */
  noLayout: PropTypes.bool,
  /*
   * Responsive size for Mobile Devices
   */
  s: PropTypes.number,
  /*
   * Responsive size for Tablet Devices
   */
  m: PropTypes.number,
  /*
   * Responsive size for Desktop Devices
   */
  l: PropTypes.number,
  /**
   *  Responsive size for Large Desktop Devices
   */
  xl: PropTypes.number,
  /*
   * disabled input
   */
  disabled: PropTypes.bool,
  /*
   * Placeholder string
   */
  placeholder: PropTypes.string,
  /*
   * override id
   * @default idgen()
   */
  id: PropTypes.string,
  /*
   * prefix icon
   */
  icon: PropTypes.string,
  /*
   * label text
   */
  label: PropTypes.string,
   /*
   * Input initial value
   */
  defaultValue: PropTypes.string,
  /*
   * Input value
   */
  value: PropTypes.string,
  /*
   * Add validate class to input
   */
  validate: PropTypes.bool,
  /*
   * Custom success message
   */
  success: PropTypes.string,
  /*
   * Custom error message
   */
  error: PropTypes.string,
  /*
   * Additional classes for input
   */
  inputClassName: PropTypes.string,
  /*
   * override type="text"
   */
  type: PropTypes.string,
  /*
   * onChange callback
   */
  onChange: PropTypes.func,
  /*
   * password type
   */
  password: PropTypes.bool,
  /*
   * email type
   */
  email: PropTypes.bool
};

export default TextInput;
