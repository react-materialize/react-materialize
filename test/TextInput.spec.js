import React from 'react';
// import { shallow, mount } from 'enzyme';
import TextInput from '../src/TextInput';
import Icon from '../src/Icon';

describe.skip('<TextInput />', () => {
  let wrapper;

  test('handles layout classes', () => {
    wrapper = shallow(<TextInput s={4} m={6} />);

    expect(wrapper.hasClass('input-field')).toBeTruthy();
    expect(wrapper.hasClass('col')).toBeTruthy();
    expect(wrapper.hasClass('s4')).toBeTruthy();
    expect(wrapper.hasClass('m6')).toBeTruthy();
  });

  test('handles removing layout classes', () => {
    wrapper = shallow(<TextInput s={4} m={6} noLayout />);

    expect(wrapper.hasClass('input-field')).toBeTruthy();
    expect(wrapper.hasClass('col')).toBeFalsy();
    expect(wrapper.hasClass('s4')).toBeFalsy();
    expect(wrapper.hasClass('m6')).toBeFalsy();
  });

  test('renders placeholder', () => {
    wrapper = shallow(<TextInput placeholder="Name" />);
    expect(wrapper.find('input').prop('placeholder')).toEqual('Name');
  });

  test('renders label', () => {
    wrapper = shallow(<TextInput label="Name" />);
    expect(wrapper.find('label').text()).toEqual('Name');
  });

  test('handles validate', () => {
    wrapper = shallow(<TextInput validate />);
    expect(wrapper.find('input.validate')).toHaveLength(1);
  });

  test('handles inputClassName', () => {
    wrapper = shallow(<TextInput validate inputClassName={'hello red'} />);
    expect(wrapper.find('input.hello.red.validate')).toHaveLength(1);
  });

  test('handles other types', () => {
    wrapper = shallow(<TextInput password />);
    expect(wrapper.find('input').prop('type')).toEqual('password');
  });

  test('handles other props', () => {
    const mockOnChange = jest.fn();
    wrapper = shallow(<TextInput onChange={mockOnChange} />);
    wrapper.find('input').simulate('change', { target: { value: '7' } });

    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });

  test('handles file type', () => {
    wrapper = shallow(<TextInput type="file" label="File" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Displays error messages', () => {
    wrapper = shallow(
      <TextInput label="Input with Error" error="this is an error" />
    );
    expect(wrapper.find('input').hasClass('invalid')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  describe('Icon', () => {
    test('renders as a string', () => {
      wrapper = shallow(<TextInput icon="cloud" />);
      expect(wrapper.find('i.material-icons.prefix')).toHaveLength(1);
    });

    test('renders as a node', () => {
      const mockIconClick = jest.fn();
      wrapper = shallow(
        <TextInput icon={<Icon onClick={mockIconClick}>cloud</Icon>} />
      );

      const iconWrapper = wrapper.find('Icon').shallow();
      expect(iconWrapper.find('i.material-icons.prefix')).toHaveLength(1);

      wrapper.find('Icon').simulate('click');
      expect(mockIconClick).toBeCalled();
    });
  });
});
