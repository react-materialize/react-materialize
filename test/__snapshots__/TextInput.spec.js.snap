// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> Displays error messages 1`] = `
<div
  className="input-field col"
>
  <input
    className="invalid"
    id="TextInput-mockId"
    type="text"
  />
  <label
    className=""
    data-error="this is an error"
    htmlFor="TextInput-mockId"
  >
    Input with Error
  </label>
  <span
    className="helper-text"
    data-error="this is an error"
  />
</div>
`;

exports[`<TextInput /> handles file type 1`] = `
<div
  className="input-field col file-field"
>
  <div
    className="btn"
  >
    <span>
      File
    </span>
    <input
      className=""
      id="TextInput-mockId"
      type="file"
    />
  </div>
  <div
    className="file-path-wrapper"
  >
    <input
      className="file-path validate"
      type="text"
    />
  </div>
</div>
`;
